### PR TITLE
Added GeoJSON export for single trees

### PR DESCRIPTION
**Pull request by ankitecd**

### CODE REVIEW

**Review by tejassonar — APPROVED** [review]

**Comment by norbertschuler** [comment]
I do not know the rules for the GeoJSON export, but when we transfer data to the backend API our data is structured differently from export. Maybe we want to make it more consistent?, e.g. regarding using a `metadata` property or what's grouped into the `app` property?

API:

```JSON
{
  "type": "single",
  "captureMode": "on-site",
  "deviceLocation": {
    "coordinates": [
      9.935418322225985,
      53.547927406691315
    ],
    "type": "Point"
  },
  "geometry": {
    "coordinates": [
      9.935418322225985,
      53.547927406691315
    ],
    "type": "Point"
  },
  "plantDate": "2021-08-05",
  "registrationDate": "2021-08-05",
  "metadata": {
    "public": {
      "Tester": "Norbert"
    },
    "private": {},
    "app": {
      "deviceLocation": [
        53.547927406691315,
        9.935418322225985
      ],
      "appVersion": "1.0.4",
      "deviceBrand": "Apple",
      "deviceModel": "iPhone 11",
      "deviceSystemName": "iOS",
      "deviceSystemVersion": "14.5",
      "deviceManufacturer": "Apple"
    }
  },
  "measurements": {
    "height": 4,
    "width": 3
  },
  "scientificSpecies": "sspec_i08DZUvNbyL0alNuEq",
  "tag": "test"
}
```

Export:

```JSON
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "isPolygonComplete": true,
        "public": {
          "Tester": "Norbert"
        },
        "private": {
        },
        "app": {
          "appVersion": "1.0.4",
          "treeType": "single",
          "captureMode": "on-site",
          "deviceBrand": "Apple",
          "deviceModel": "iPhone 11",
          "deviceManufacturer": "Apple",
          "deviceSystemName": "iOS",
          "deviceSystemVersion": "14.5",
          "speciesHeight": 4,
          "speciesDiameter": 3,
          "tagId": "test",
          "plantationDate": "2021-08-05T11:41:13.469Z",
          "species": [
            {
              "aliases": "Baum antsingyae",
              "treeCount": 1,
              "id": "sspec_i08DZUvNbyL0alNuEq"
            }
          ],
          "deviceLocation": [
            53.547927406691315,
            9.935418322225985
          ]
        }
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          9.935418322225985,
          53.547927406691315
        ]
      }
    }
  ]
}
```

**Comment by tejassonar** [comment]
> @tejassonar I do not know the rules for the GeoJSON export, but when we transfer data to the backend API our data is structured differently from exporting it. Maybe we want to make it more consistent, e.g. regarding using a `metadata` property or what's grouped into the `app` property? (If we do any changes, we would have to apply them to multiple trees registration export as well of course.)
> 
> API:
> 
> ```json
> {
>   "type": "single",
>   "captureMode": "on-site",
>   "deviceLocation": {
>     "coordinates": [
>       9.935418322225985,
>       53.547927406691315
>     ],
>     "type": "Point"
>   },
>   "geometry": {
>     "coordinates": [
>       9.935418322225985,
>       53.547927406691315
>     ],
>     "type": "Point"
>   },
>   "plantDate": "2021-08-05",
>   "registrationDate": "2021-08-05",
>   "metadata": {
>     "public": {
>       "Tester": "Norbert"
>     },
>     "private": {},
>     "app": {
>       "deviceLocation": [
>         53.547927406691315,
>         9.935418322225985
>       ],
>       "appVersion": "1.0.4",
>       "deviceBrand": "Apple",
>       "deviceModel": "iPhone 11",
>       "deviceSystemName": "iOS",
>       "deviceSystemVersion": "14.5",
>       "deviceManufacturer": "Apple"
>     }
>   },
>   "measurements": {
>     "height": 4,
>     "width": 3
>   },
>   "scientificSpecies": "sspec_i08DZUvNbyL0alNuEq",
>   "tag": "test"
> }
> ```
> 
> Export:
> 
> ```json
> {
>   "type": "FeatureCollection",
>   "features": [
>     {
>       "type": "Feature",
>       "properties": {
>         "isPolygonComplete": true,
>         "public": {
>           "Tester": "Norbert"
>         },
>         "private": {
>         },
>         "app": {
>           "appVersion": "1.0.4",
>           "treeType": "single",
>           "captureMode": "on-site",
>           "deviceBrand": "Apple",
>           "deviceModel": "iPhone 11",
>           "deviceManufacturer": "Apple",
>           "deviceSystemName": "iOS",
>           "deviceSystemVersion": "14.5",
>           "speciesHeight": 4,
>           "speciesDiameter": 3,
>           "tagId": "test",
>           "plantationDate": "2021-08-05T11:41:13.469Z",
>           "species": [
>             {
>               "aliases": "Baum antsingyae",
>               "treeCount": 1,
>               "id": "sspec_i08DZUvNbyL0alNuEq"
>             }
>           ],
>           "deviceLocation": [
>             53.547927406691315,
>             9.935418322225985
>           ]
>         }
>       },
>       "geometry": {
>         "type": "Point",
>         "coordinates": [
>           9.935418322225985,
>           53.547927406691315
>         ]
>       }
>     }
>   ]
> }
> ```

Public and Private objects include the data which is added in additional data form. All the other app data is added to the app object. Everything else is the default structure of the geoJSON format. As we did not want to repeat the data, we have kept it this way. @ankitecd Any more insights?

**Comment by norbertschuler** [comment]
> Public and Private objects include the data which is added in additional data form. All the other app data is added to the app object. Everything else is the default structure of the geoJSON format. As we did not want to repeat the data, we have kept it this way. @ankitecd Any more insights?

Why is "isPolygonComplete" and "geometry" not included into "app" then? I understood "app" is completely different from the app property in the API. I don't want to demand to make the data structure the same until both are logical and the export data format is useful and understandable structured for the user who does the export :-)

**Comment by tejassonar** [comment]
> > Public and Private objects include the data which is added in additional data form. All the other app data is added to the app object. Everything else is the default structure of the geoJSON format. As we did not want to repeat the data, we have kept it this way. @ankitecd Any more insights?
> 
> Why is "isPolygonComplete" and "geometry" not included into "app" then? I understood "app" is completely different from the app property in the API. I don't want to demand to make the data structure the same until both are logical and the export data format is useful and understandable structured for the user who does the export :-)

"isPolygonComplete" and "geometry" is also a part of geoJSON structure, I believe. And the reason why we did not add it in "app" is, it would have been duplicate data again.

**Comment by ankitecd** [comment]
> 
> Why is "isPolygonComplete" and "geometry" not included into "app" then? I understood "app" is completely different from the app property in the API. I don't want to demand to make the data structure the same until both are logical and the export data format is useful and understandable structured for the user who does the export :-)

This is the rule for GeoJSON
`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Point", // Point or Polygon
        "coordinates": [125.6, 10.1] // depends on type Array or Array of Array
      },
      "properties": { // contains extra properties
        "name": "Dinagat Islands"
      }
    }
  ]
}`

`type` and `geometry` are mandatory for GeoJSON to plot polygons or point on map. Any extra data can go in properties. 

`isPolygonComplete` was added earlier and it is not required anymore. I think we can have `measurements`, `deviceLocation`  as the API.

File: app/utils/convertInventoryToGeoJson.ts [hunk 1]
```diff
@@ -34,15 +33,16 @@ export default async function getGeoJsonData(inventoryData: any) {
...
-          type: 'LineString',
-          coordinates: onePolygon.coordinates.map((oneCoordinate: any) => [
-            oneCoordinate.longitude,
-            oneCoordinate.latitude,
-          ]),
+          type: 'Polygon',
+          coordinates: [
+            onePolygon.coordinates.map((oneCoordinate: any) => [
+              oneCoordinate.longitude,
+              oneCoordinate.latitude,
```
Comment (norbertschuler) [inline app/utils/convertInventoryToGeoJson.ts:43]: You wrote longitude before latitude here in contrast to all other places where latitude is listed first. Is this an error?

Comment (norbertschuler) [reply app/utils/convertInventoryToGeoJson.ts]: Fixed it myself.

**Review by norbertschuler — APPROVED** [review]
Looks good for me.

File: app/utils/convertInventoryToGeoJson.ts (hunk 1 above)
Comment (ankitecd) [reply app/utils/convertInventoryToGeoJson.ts]: I was testing geoJSON on this website [https://geojson.io](https://geojson.io) and found out that for point [lat, long] is correct and for Polygon it should be [long, lat]

Comment (ankitecd) [reply app/utils/convertInventoryToGeoJson.ts]: You can check by exporting you geoJSON

Comment (ankitecd) [reply app/utils/convertInventoryToGeoJson.ts]: I think there's some other problem, I'm looking into this

Comment (norbertschuler) [reply app/utils/convertInventoryToGeoJson.ts]: From my tests with a single tree and multiple tree export on this website, we seem to need [longitude, latitude] for single point as well as polygons, e.g.
```
  "geometry": {
    "coordinates": [
      9.935418322225985,
      53.547927406691315
    ],
    "type": "Point"
  },
```
```
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              9.9353538196508,
              53.54860199113149
            ],
            [
              9.93557510185127,
              53.54860039762337
            ],
            [
              9.935572419770125,
              53.54870716926696
            ],
            [
              9.935352478590488,
              53.548703982105074
            ],
            [
              9.9353538196508,
              53.54860199113149
            ]
          ]
        ]
      }
```

But I am irritated looking at above API example why our device location defined it differently.
```
  "deviceLocation": {
    "coordinates": [
      9.935418322225985,
      53.547927406691315
    ],
    "type": "Point"
  },
```

We should really check this thoroughly :-)

Comment (ankitecd) [reply app/utils/convertInventoryToGeoJson.ts]: Yes, I'm debugging the issue and so converted this PR to Draft.

**Review by norbertschuler — CHANGES_REQUESTED** [review]
I made several test exports with new and existing registrations. It looks ok for me except the fact that the `deviceLocation` for sample tree got truncated to 5 digits not only in the UI but also in the export (you did not truncate the deviceLocation for the whole multiple tree registration, but only for sample trees - both in the UI and GeoJSON export):

```JSON
          "deviceLocation": {
            "coordinates": [
              9.9355,
              53.5485
            ],
            "type": "Point"
          },

```
And for some unknown reason the backend API failed to accept this data with the following error as well:  
`"Error in one or more bulk request actions:\n\nupdate: /heroku-development.plant_location/_doc/409 caused failed to parse field [samplePlantLocations.metadata.app.deviceLocation] of type [float] in document with id '409'. Preview of field's value: '{coordinates=[53.5485, 9.9355], type=Point}'\n\nfile: /app/vendor/ruflin/elastica/src/Bulk.php, line: 339"
`

[TestExports.zip](https://github.com/Plant-for-the-Planet-org/treemapper/files/6973442/TestExports.zip)

**Comment by norbertschuler** [comment]
> I made several test exports with new and existing registrations. It looks ok for me except the fact that the `deviceLocation` for sample tree got truncated to 5 digits not only in the UI but also in the export (you did not truncate the deviceLocation for the whole multiple tree registration, but only for sample trees - both in the UI and GeoJSON export):
> 
> ```json
>           "deviceLocation": {
>             "coordinates": [
>               9.9355,
>               53.5485
>             ],
>             "type": "Point"
>           },
> ```

We do not fix this now.

> And for some unknown reason the backend API failed to accept this data with the following error as well:
> `"Error in one or more bulk request actions:\n\nupdate: /heroku-development.plant_location/_doc/409 caused failed to parse field [samplePlantLocations.metadata.app.deviceLocation] of type [float] in document with id '409'. Preview of field's value: '{coordinates=[53.5485, 9.9355], type=Point}'\n\nfile: /app/vendor/ruflin/elastica/src/Bulk.php, line: 339" `

This got fixed on develop and I do not think this bug exists on production yet.

File: app/utils/additionalData/functions.ts [hunk 2]
```diff
@@ -321,10 +321,16 @@ export const basicAppAdditionalDataForAPI = ({ data, isSampleTree = false }: IGe
...
+        coordinates: [coords[0].latitude, coords[0].longitude],
+        type: 'Point',
+      };
     }
   } else {
-    appAdditionalDetails['deviceLocation'] = [data.deviceLatitude, data.deviceLongitude];
+    appAdditionalDetails['deviceLocation'] = {
+      coordinates: [data.deviceLatitude, data.deviceLongitude],
+      type: 'Point',
+    };
```
[the remark this answers is not in the document]
Comment (norbertschuler) [reply app/utils/additionalData/functions.ts]: I want to test this once more…

**Review by norbertschuler — APPROVED** [review]
I created registrations for all four different types and checked the GeoJSON export as well as the upload with API calls. Everywhere the order is now [long, lat].

[TestExports.zip](https://github.com/Plant-for-the-Planet-org/treemapper/files/6975613/TestExports.zip)

**Pull request merged**